### PR TITLE
Revise compiler hardening options on Unix

### PR DESCRIPTION
**Pull request by am11**
* stack-protector-strong
* stack-clash-protection
* cf-protection
* fortification

This is an attempt to harden the runtime builds based on https://developers.redhat.com/articles/2022/06/02/use-compiler-flags-stack-protection-gcc-and-clang and similar articles/practices, e.g. Firefox builds use these options (modulo control-flow protection and fortify level 2 instead of 3), but at the same time, they provide an umbrella option to bail out of hardening mode. Currently we enable strong stack protection and allow fortification set by the build environment in release configuration. This PR is hardening the defaults and introducing two more options to increase program security.

### CODE REVIEW

File: eng/native/configurecompiler.cmake [hunk 1]
```diff
@@ -532,20 +532,36 @@ if (CLR_CMAKE_HOST_UNIX)
...
     check_c_compiler_flag(-fstack-protector-strong COMPILER_SUPPORTS_F_STACK_PROTECTOR_STRONG)
     if (COMPILER_SUPPORTS_F_STACK_PROTECTOR_STRONG)
       add_compile_options(-fstack-protector-strong)
     endif()
-  endif(CLR_CMAKE_HOST_OSX OR CLR_CMAKE_HOST_MACCATALYST)
+
+    check_c_compiler_flag(-fstack-clash-protection COMPILER_SUPPORTS_F_STACK_CLASH_PROTECTION)
+    # explicit check for gcc is required because while clang does not implement stack-protection,
+    # it does not return 'unknown argument' either, instead:
+    # clang: warning: argument unused during compilation: '-fstack-clash-protection' [-Wunused-command-line-argument]
```
Comment (am11) [inline eng/native/configurecompiler.cmake:551]:
```suggestion
    #   clang: warning: argument unused during compilation: '-fstack-clash-protection' [-Wunused-command-line-argument]
    # see:
    #   https://github.com/llvm/llvm-project/issues/40148
    #   https://gitlab.kitware.com/cmake/cmake/-/issues/25390
```

File: eng/native/configurecompiler.cmake [hunk 2]
```diff
@@ -532,20 +532,39 @@ if (CLR_CMAKE_HOST_UNIX)
...
+      add_compile_options(-fstack-clash-protection)
+    endif()
+
+    check_c_compiler_flag(-fcf-protection COMPILER_SUPPORTS_F_CONTROL_FLOW_PROTECTION)
+    if (COMPILER_SUPPORTS_F_CONTROL_FLOW_PROTECTION)
+      add_compile_options(-fcf-protection)
+    endif()
+
+    # Enable maximum fortification
+    remove_definitions(-D_FORTIFY_SOURCE)
```
Comment (am11) [inline eng/native/configurecompiler.cmake:565]:
```suggestion
    string(REGEX REPLACE "-D_FORTIFY_SOURCE=[0-9] " "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
    string(REGEX REPLACE "-D_FORTIFY_SOURCE=[0-9] " "" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
```

**Comment by jkotas** [comment]
Are any of these options implemented by clang or is this just changing the options for gcc?

File: eng/native/configurecompiler.cmake [hunk 3]
```diff
@@ -532,20 +532,40 @@ if (CLR_CMAKE_HOST_UNIX)
...
+    # see:
+    #   https://github.com/llvm/llvm-project/issues/40148
+    #   https://gitlab.kitware.com/cmake/cmake/-/issues/25390
+    if (COMPILER_SUPPORTS_F_STACK_CLASH_PROTECTION AND CMAKE_C_COMPILER_ID STREQUAL "GNU")
+      add_compile_options(-fstack-clash-protection)
+    endif()
+
+    check_c_compiler_flag(-fcf-protection COMPILER_SUPPORTS_F_CONTROL_FLOW_PROTECTION)
+    if (COMPILER_SUPPORTS_F_CONTROL_FLOW_PROTECTION)
+      add_compile_options(-fcf-protection)
```
Comment (jkotas) [inline eng/native/configurecompiler.cmake:561]: Are Linux distros compiled with this turned on? The control flow protection is meaningful only if it is implemented by all code in the process.

**Comment by am11** [comment]
> Are any of these options implemented by clang or is this just changing the options for gcc?

Only stack-clash-protection is not implemented by clang, rest of them are supported by both clang and gcc.

**Comment by jkotas** [comment]
> rest of them are supported by both clang and gcc.

Any ideas what is the perf overhead of all these protections that you are trying to enable?

**Comment by am11** [comment]
I'll try to measure. Any suggestions; e.g. a few APIs which are critical? There is a theoretical perf. overhead due to fortification (per https://fedoraproject.org/wiki/Changes/Add_FORTIFY_SOURCE%3D3_to_distribution_build_flags#Performance_note), Since glibc-based distros are enabling it (it's a [no-op](https://wiki.musl-libc.org/future-ideas) on musl-libc), we can align with them? Alternative is to leave it as-is (distro defaults, which we have right now). We can also add them behind a flag, e.g. `build.sh --hardened`.

**Comment by am11** [comment]
Size diffs (percentage increase) on linux-arm64 are mixed:

```sh
# before / main / HEAD~4
$ git checkout HEAD~4
$ rm -rf artifacts
$ ./build.sh clr+libs -c Release
$ find artifacts/bin -name *.so -exec du -b {} \; > /tmp/before-du

# after / pr
$ git checkout feature/compiler-hardening-unix
$ rm -rf artifacts
$ ./build.sh clr+libs -c Release
$ find artifacts/bin -name *.so -exec du -b {} \; > /tmp/after-du

$ while read -r size1 name1 && read -r size2 name2 <&3; do
      PERC=$(echo "scale=3; ($size2 - $size1)/ $size1 * 100" | bc | sed "s/^\(-\?\)\./\10./")
      echo "$PERC% $name1"
  done < /tmp/before-du 3< /tmp/after-du |
  grep -v '^0%'

0.100% artifacts/bin/crossgen-corelib/arm64/Release/libclrjit_universal_arm64_arm64.so
0.100% artifacts/bin/crossgen-corelib/arm64/Release/libclrjit_win_x86_arm64.so
0.100% artifacts/bin/crossgen-corelib/arm64/Release/libclrjit_universal_arm_arm64.so
0.100% artifacts/bin/crossgen-corelib/arm64/Release/libclrjit_win_x64_arm64.so
0.100% artifacts/bin/crossgen-corelib/arm64/Release/libclrjit_unix_x64_arm64.so
-4.300% artifacts/bin/native/net9.0-linux-Release-arm64/libSystem.Security.Cryptography.Native.OpenSsl.so
6.400% artifacts/bin/native/net9.0-linux-Release-arm64/libSystem.Globalization.Native.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/libclrjit_universal_arm64_arm64.so
0.300% artifacts/bin/coreclr/linux.arm64.Release/sharedFramework/libmscordaccore.so
0.300% artifacts/bin/coreclr/linux.arm64.Release/sharedFramework/libclrjit.so
2.000% artifacts/bin/coreclr/linux.arm64.Release/sharedFramework/libcoreclrtraceptprovider.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/sharedFramework/libcoreclr.so
-0.500% artifacts/bin/coreclr/linux.arm64.Release/sharedFramework/libclrgc.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/libclrjit_win_x86_arm64.so
0.300% artifacts/bin/coreclr/linux.arm64.Release/libmscordaccore.so
0.300% artifacts/bin/coreclr/linux.arm64.Release/libclrjit.so
0.300% artifacts/bin/coreclr/linux.arm64.Release/libsuperpmi-shim-simple.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/libclrjit_universal_arm_arm64.so
2.000% artifacts/bin/coreclr/linux.arm64.Release/libcoreclrtraceptprovider.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/libclrjit_win_x64_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/libclrjit_unix_x64_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/crossgen2/libclrjit_universal_arm64_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/crossgen2/libclrjit_win_x86_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/crossgen2/libclrjit_universal_arm_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/crossgen2/libclrjit_win_x64_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/crossgen2/libclrjit_unix_x64_arm64.so
0.300% artifacts/bin/coreclr/linux.arm64.Release/libsuperpmi-shim-counter.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/libcoreclr.so
-0.500% artifacts/bin/coreclr/linux.arm64.Release/libclrgc.so
0.300% artifacts/bin/coreclr/linux.arm64.Release/libsuperpmi-shim-collector.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/arm64/crossgen2/tools/libclrjit_universal_arm64_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/arm64/crossgen2/tools/libclrjit_win_x86_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/arm64/crossgen2/tools/libclrjit_universal_arm_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/arm64/crossgen2/tools/libclrjit_win_x64_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/arm64/crossgen2/tools/libclrjit_unix_x64_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/ilc-published/libclrjit_universal_arm64_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/ilc-published/libclrjit_win_x86_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/ilc-published/libclrjit_universal_arm_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/ilc-published/libclrjit_win_x64_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/ilc-published/libclrjit_unix_x64_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/ilc/libclrjit_universal_arm64_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/ilc/libclrjit_win_x86_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/ilc/libclrjit_universal_arm_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/ilc/libclrjit_win_x64_arm64.so
0.100% artifacts/bin/coreclr/linux.arm64.Release/ilc/libclrjit_unix_x64_arm64.so
```
(and 37 out of 119 binaries unchanged `grep -v '^0%'` -> `grep '^0%'`)

File: eng/native/configureplatform.cmake [hunk 4]
```diff
@@ -492,10 +492,3 @@ if(NOT CLR_CMAKE_TARGET_BROWSER AND NOT CLR_CMAKE_TARGET_WASI)
 
     set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 endif()
-
-string(TOLOWER "${CMAKE_BUILD_TYPE}" LOWERCASE_CMAKE_BUILD_TYPE)
-if(LOWERCASE_CMAKE_BUILD_TYPE STREQUAL debug)
-    # Clear _FORTIFY_SOURCE=2, if set
```
Comment (jkotas) [inline eng/native/configureplatform.cmake:498]: Why was this needed and why is it not necessary anymore?

File: eng/native/configurecompiler.cmake [hunk 5]
```diff
@@ -532,20 +532,40 @@ if (CLR_CMAKE_HOST_UNIX)
...
+    endif()
+
+    # Enable maximum fortification
+    string(REGEX REPLACE "-D_FORTIFY_SOURCE=[0-9] " "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
```
Comment (jkotas) [inline eng/native/configurecompiler.cmake:565]: Why do we need to throw away what was passed in? It does not look right to ignore conflicting settings silently.

File: eng/native/configurecompiler.cmake [hunk 6]
```diff
@@ -532,20 +532,40 @@ if (CLR_CMAKE_HOST_UNIX)
...
+  # hardening options
+  if(NOT CLR_CMAKE_HOST_BROWSER AND NOT CLR_CMAKE_HOST_WASI)
     check_c_compiler_flag(-fstack-protector-strong COMPILER_SUPPORTS_F_STACK_PROTECTOR_STRONG)
     if (COMPILER_SUPPORTS_F_STACK_PROTECTOR_STRONG)
       add_compile_options(-fstack-protector-strong)
     endif()
-  endif(CLR_CMAKE_HOST_OSX OR CLR_CMAKE_HOST_MACCATALYST)
+
+    check_c_compiler_flag(-fstack-clash-protection COMPILER_SUPPORTS_F_STACK_CLASH_PROTECTION)
+    # explicit check for gcc is required because while clang does not implement stack-protection,
```
Comment (jkotas) [inline eng/native/configurecompiler.cmake:549]: https://github.com/llvm/llvm-project/issues/40148 says that clang implements it on x86 architectures. Should we enable it on those at least?

**Comment by jkotas** [comment]
> -0.500% artifacts/bin/coreclr/linux.arm64.Release/sharedFramework/libclrgc.so

It is unexpected for the binary size to decrease with the extra hardening. What is going on?

**Comment by jkotas** [comment]
> I'll try to measure. Any suggestions; e.g. a few APIs which are critical? 

I think we should break this into one PR per option, so that we can reason about each separately and make it easier to be able to track down potential regressions.

- I am not worried about `fstack-clash-protection`. (I am actually surprised that it is not enabled yet.)
- `fcf-protection` adds an extra instruction in many places. I think we would want to run JIT and/or GC throughput tests to see the impact.
- `D_FORTIFY_SOURCE` adds defense-in-depth error handling around number of C-runtime functions, including very primitive methods like memcpy and memset. It has a risk of having outliers where this extra error handling gets added to on a hot path.

File: eng/native/configurecompiler.cmake (hunk 5 above)
Comment (am11) [reply eng/native/configurecompiler.cmake:565]: It is not new, we moved it from configureplatform.cmake.

Comment (jkoritzinsky) [reply eng/native/configurecompiler.cmake:565]: Instead of throwing away the options, do we want to use the existing option if it's already specified by the platform build scripts?

I've noticed that some community distro builds have scripts to manipulate these flags because their default settings conflict with ours.

Comment (am11) [reply eng/native/configurecompiler.cmake:565]: Yes, I think it will be better to honor the distro choice here. We can scan the `CMAKE_C{XX}_FLAGS` for `_FORTIFY_SOURCE` and set the default value only when it is not set by caller.

File: eng/native/configureplatform.cmake (hunk 4 above)
Comment (am11) [reply eng/native/configureplatform.cmake:498]: This was done for Clear Linux OS, which sets `_FORTIFY_SOURCE=2` globally.

From the [article](https://developers.redhat.com/articles/2022/06/02/use-compiler-flags-stack-protection-gcc-and-clang#)
> GCC and Clang select fortified sources using the preprocessor flag `-D_FORTIFY_SOURCE=<n>` with an optimization level greater than `-O0`. A higher <n> corresponds to greater protection. However, there may be a tradeoff in performance, and some programs that conform to the standard may fail due to stricter security checks. GCC ([version 12](https://developers.redhat.com/articles/2022/04/25/new-c-features-gcc-12) and later) and Clang (version 9.0 and later) support `<n>` up to 3.

The debug build (`-O0`) on Clear Linux was failing: https://github.com/dotnet/runtime/issues/2188.

This PR is moving this to configurecompiler.cmake.

File: src/tests/Common/Platform/platformdefines.cpp [hunk 7]
```diff
@@ -321,7 +321,7 @@ DWORD TP_GetFullPathName(LPWSTR fileName, DWORD nBufferLength, LPWSTR lpBuffer)
     return GetFullPathNameW(fileName, nBufferLength, lpBuffer, NULL);
 #else
     char nativeFullPath[MAX_PATH];
-    (void)realpath(HackyConvertToSTR(fileName), nativeFullPath);
+    (void)!realpath(HackyConvertToSTR(fileName), nativeFullPath);
```
Comment (AaronRobinsonMSFT) [inline src/tests/Common/Platform/platformdefines.cpp:324]: What does the `!` provide?

Comment (am11) [reply src/tests/Common/Platform/platformdefines.cpp:324]: It's to suppress the unused result warning with gcc for the APIs decorated with `__attribute__((warn_unused_result))`. Most people found it unexpected and counter-intuitive: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66425; especially when the similar spec of C++17's `[[nodiscard]]` explicitly allow [void-to-cast suppression](https://timsong-cpp.github.io/cppwp/n4659/dcl.attr#nodiscard-2) and gcc abides by it, but with respect to WUR, there is no standard in C so they made this choice to insist of warning with void.

BTW, we use in a few places as-is:
```
src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:    (void)!write(STDERR_FILENO, message, strlen(message));
src/coreclr/pal/src/exception/signal.cpp:                (void)!write(STDERR_FILENO, StackOverflowMessage, sizeof(StackOverflowMessage) - 1);
src/coreclr/vm/encee.cpp:        (void)!alloca(frameIncrement);
src/native/libs/System.Native/pal_process.c:    (void)! pipe2(waitForChildToExecPipe, O_CLOEXEC);
src/native/libs/System.Native/pal_process.c:    (void)! SystemNative_Pipe(waitForChildToExecPipe, PAL_O_CLOEXEC);
```

Comment (am11) [reply src/tests/Common/Platform/platformdefines.cpp:324]: It's to suppress the unused result warning with gcc for the APIs decorated with `__attribute__((warn_unused_result))`. Most people found it unexpected and counter-intuitive: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66425; especially when the similar spec of C++17's `[[nodiscard]]` explicitly allow [cast-to-void suppression](https://timsong-cpp.github.io/cppwp/n4659/dcl.attr#nodiscard-2) and gcc abides by it, but with respect to WUR, there is no standard in C so they made this choice to insist of warning with void.

BTW, we use this in a few places as-is:
```sh
src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:754:    (void)!write(STDERR_FILENO, message, strlen(message));
src/coreclr/pal/src/exception/signal.cpp:600:                (void)!write(STDERR_FILENO, StackOverflowMessage, sizeof(StackOverflowMessage) - 1);
src/coreclr/vm/encee.cpp:712:        (void)!alloca(frameIncrement);
src/native/libs/System.Native/pal_process.c:307:    (void)! pipe2(waitForChildToExecPipe, O_CLOEXEC);
src/native/libs/System.Native/pal_process.c:309:    (void)! SystemNative_Pipe(waitForChildToExecPipe, PAL_O_CLOEXEC);
```

Comment (AaronRobinsonMSFT) [reply src/tests/Common/Platform/platformdefines.cpp:324]: > It's to suppress the unused result warning with gcc for the APIs decorated with __attribute__((warn_unused_result)). 

Interesting. I didn't know that was a thing. I've been using `(void)` for years and that was the indication I didn't care. I agree it is a bit unexpected. Thanks for the link.
